Stats: Add new commercial purchase page

grzegorz-cp opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #80259

## Proposed Changes

* introduce a new purchase page for a single plan (not a wizard like previously)
* combine the single-page design with the existing commercial plan purchase flow
* create a new file with shared components
* extended pages visited by owners of a free or PWYW plans with a notice above the wizard

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* navigate to the live branch
* without any feature flag applied verify that the experience didn't change for:
   * a page without any plan purchased (you should see the old wizard)
   * a page with the commercial plan purchased (you should see a confirmation notice)
   * a page with free or PWYW plan purchased (you should see the old wizard)
* apply `stats/type-detection` feature flag and check the updated experience
  * if the page wasn't classified as either personal or commercial, you should see the old wizard
  * if your page has been classified as commercial you should see a new purchase page for only the commercial plan (not the wizard) - this can be tested locally by forcing commercial detection in the controller (`isCommercial` option)
  * if your page has been classified as personal you should see the wizard (a separate page is currently in progress)
  * if you purchased free or PWYW plans you should see a banner on top the wizard when visiting the page again

New commercial page:
<img width="1059" alt="SCR-20230831-nfhc" src="https://github.com/Automattic/wp-calypso/assets/112354940/67e455a1-e82d-48c1-8d2c-f6eb541d4012">

Additional pages available when visiting the purchase page with a free or personal plans:
<img width="1187" alt="SCR-20230901-obxg" src="https://github.com/Automattic/wp-calypso/assets/112354940/7d386ad8-acfb-4f6f-b847-b7c82f3f347c">

<img width="1188" alt="SCR-20230901-ocbn" src="https://github.com/Automattic/wp-calypso/assets/112354940/a23de73c-63e9-4440-84cc-d3ae3c700753">


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

grzegorz-cp committed:
Stats: Add new commercial purchase page

grzegorz-cp committed:
Stats: Adding notice banners for personal plan owners

grzegorz-cp committed:
Merge branch 'trunk' into add/stats-80259-commercial-flow

dognose24 commented on client/my-sites/stats/stats-purchase/index.tsx:
```diff
@@ -173,6 +178,72 @@ const StatsPurchasePage = ( {
+						}
+						{
+							// blog doesn't have any plan but is not categorised as either personal or commectial - show old purchase wizard
+							( isFreeOwned || isPWYWOwned || ( showPurchasPage && ! showSinglePurchasPage ) ) && (
```
```suggestion
							( isFreeOwned || isPWYWOwned || options.isCommercial === null ) && (
```

This condition `( showPurchasPage && ! showSinglePurchasPage )` could be simplified to `options.isCommercial === null`, which means the site has not been classified.

dognose24 replied:
The wizard shows forever if I have a free or PWYW plan, no matter what value of `options.isCommercial` is, even when I already have bought a commercial plan.

<img width="1290" alt="buy_commercial_but_is_not_determined" src="https://github.com/Automattic/wp-calypso/assets/6869813/3d7f009a-01ed-4cd9-958f-e2660ae0b0a9">

dognose24 commented on client/my-sites/stats/stats-purchase/index.tsx:
```diff
@@ -173,6 +178,72 @@ const StatsPurchasePage = ( {
+								/>
+							)
+						}
+						{ showPurchasPage && showSinglePurchasPage && (
```
```suggestion
						{ showSinglePurchasPage && (
```

We don't need to check `showPurchasPage` here since `showSinglePurchasPage` has included the checking.

dognose24 commented on client/my-sites/stats/stats-purchase/index.tsx:
```diff
@@ -126,6 +128,9 @@ const StatsPurchasePage = ( {
 
 	const maxSliderPrice = commercialMonthlyProduct?.cost;
 
+	const showPurchasPage = ! isCommercialOwned && ! isFreeOwned && ! isPWYWOwned;
```
```suggestion
	const showPurchasePage = ! isCommercialOwned && ! isFreeOwned && ! isPWYWOwned;
```

kangzj replied:
fixed in c532ddde7c

dognose24 commented on client/my-sites/stats/stats-purchase/index.tsx:
```diff
@@ -126,6 +128,9 @@ const StatsPurchasePage = ( {
 	const maxSliderPrice = commercialMonthlyProduct?.cost;
 
+	const showPurchasPage = ! isCommercialOwned && ! isFreeOwned && ! isPWYWOwned;
+	const showSinglePurchasPage = options.isCommercial !== null && showPurchasPage; // blog has been already categorised as either personal or commercial and doesn't have a plan purchased
```
```suggestion
	const showSinglePurchasePage = options.isCommercial !== null && showPurchasPage; // blog has been already categorised as either personal or commercial and doesn't have a plan purchased
```

dognose24 commented on client/my-sites/stats/stats-purchase/stats-purchase-notice.jsx:
```diff
@@ -27,44 +25,68 @@ const StatsCommercialOwned = ( { siteSlug } ) => {
-				</div>
-			</Card>
+		<div className="stats-purchase-page__notice">
+			{ isCommercialOwned && <StatsPurchaseNotice siteSlug={ siteSlug } /> }
```
We may need to apply the `StatsNotices` [system](https://github.com/Automattic/wp-calypso/blob/add/stats-80259-commercial-flow/client/my-sites/stats/stats-notices/index.tsx#L28-L41) to avoid showing all the notices on the same page.

dognose24 reviewed (commented):
For testing applying the feature flag `stats/type-detection`, I tested for a site marked as commercial with a free or PWYW plan purchased. I should go to the `StatsSingleItemPagePurchase`, but it still shows the wizard because the `showPurchasPage` is false. There would be some condition conflicts between owned plans and `options.isCommercial` checking.

grzegorz-cp committed:
Merge branch 'trunk' into add/stats-80259-commercial-flow

grzegorz-cp committed:
Stats: Adding redirects for upgrading

grzegorz-cp commented:
Thank you for the review @dognose24. I removed the notice banners and replaced them with an entire page, like when having a commercial and revisiting the purchase page. 
I added redirects for upgrading and updated the logic, showing different components.

kangzj committed:
fix typo

kangzj commented on client/my-sites/stats/stats-purchase/stats-purchase-notice.jsx:
```diff
@@ -27,44 +63,126 @@ const StatsCommercialOwned = ( { siteSlug } ) => {
+				{ translate( 'See your stats' ) }
+			</Button>
+
+			<Button
```
What do you think if we reduce the scope of the PR a bit by removing the upgrade button and the relevant code? Because it doesn't really work anyway.

kangzj commented on client/my-sites/stats/stats-purchase/stats-purchase-notice.jsx:
```diff
@@ -27,44 +63,126 @@ const StatsCommercialOwned = ( { siteSlug } ) => {
+			<Button variant="primary" onClick={ handleClick }>
+				{ translate( 'See your stats' ) }
+			</Button>
+			<Button
```
What do you think if we reduce the scope of the PR a bit by removing the upgrade button and the relevant code? Because it doesn't really work anyway.

kangzj commented on client/my-sites/stats/stats-purchase/stats-purchase-notice.jsx:
```diff
@@ -27,44 +63,126 @@ const StatsCommercialOwned = ( { siteSlug } ) => {
-				</div>
-			</Card>
+		<div className="stats-purchase-page__notice">
+			{ isCommercialOwned && <StatsPurchaseNotice siteSlug={ siteSlug } /> }
```
I guess the notices wouldn't be required anymore per the design, would they?

dognose24 replied:
I see a notice informing users who have purchased the `commercial` plan. But I didn't see notices for `free` and `personal` plans on the wrap-up design either. 🤔

<img width="912" alt="截圖 2023-09-02 上午12 31 34" src="https://github.com/Automattic/wp-calypso/assets/6869813/075a3033-d116-4ef7-8506-6e8f5e72f727">

dognose24 replied:
Two notices for `free` and `personal` plans appear simultaneously when I already have purchased both.

<img width="1255" alt="截圖 2023-09-02 上午12 32 52" src="https://github.com/Automattic/wp-calypso/assets/6869813/a195c0c5-b4c9-4251-90f7-8fda02183382">

kangzj replied:
Maybe we should only test for the highest plan?

grzegorz-cp replied:
Good idea!

kangzj reviewed (commented):
All work as described and code looks good 👍 

I'd suggest we remove the upgrade buttons (see inline comments) to reduce the scope of the PR, it's already got 3 major functionalities. 😄

dognose24 commented on client/my-sites/stats/stats-purchase/index.tsx:
```diff
@@ -173,6 +189,83 @@ const StatsPurchasePage = ( {
+												! isCommercialOwned &&
+												options.isCommercial === false ) ) && (
+											<>
+												<div>This page is being developed</div>
```
```suggestion

```

dognose24 reviewed (commented):
I agree with @kangzj's opinion of reducing the scope of this PR. The new commercial purchase page looks good to me. Since these changes are gated behind the feature flag, we can iterate for those complex conditions to show notices and redirect with owned plans and the `options.isCommercial` check.

grzegorz-cp committed:
Merge branch 'trunk' into add/stats-80259-commercial-flow

grzegorz-cp committed:
Stats: Fix multiple notice display and remove upgrade button

grzegorz-cp commented:
Thanks for the reviews @kangzj and @dognose24.
I limited the full-screen notice display to avoid information about multiple plans and commented out the upgrade button.

kangzj reviewed (approved):
It works as described and the changes look good 👍 

I'd prefer the commented code removed but which is not a blocker.

grzegorz-cp committed:
Stats: Removing comments

kangzj reviewed (approved):
:shipit:

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8861725

Thank you @grzegorz-cp for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8861725).

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8861725).
